File: 2024/Day01/Rust/Historian_Hysteria_P2.rs
use std::fs::File;
use std::io::{self, BufRead, BufReader};

fn main() -> io::Result<()> {
    let file = File::open("./src/input.txt")?;
    let reader = BufReader::new(file);
    let mut left_nums: Vec<i32> = Vec::new();
    let mut right_nums: Vec<i32> = Vec::new();
    let mut similarity_score: i32 = 0;
    for line in reader.lines() {
        match line {
            Ok(line) => {
                
                let parts: Vec<&str> =  line.split_whitespace().collect();
                left_nums.push(parts[0].parse().unwrap());
                right_nums.push(parts[1].parse().unwrap());
            },
            Err(e) => println!("Error: {}", e),
        }
    }


    for value in left_nums.iter_mut() {
        let mut score  = 0;

        for value2 in right_nums.iter_mut() {
            if value == value2 {
                score += 1;
            }
        }
        similarity_score += *value * score;
    }
    println!("{}", similarity_score);
    
    Ok(())
}

File: 2024/Day01/Python/Historian_Hysteria_P2.py
file_name = "input_Historian_Hysteria_P2.txt"

try:
    with open(file_name, "r") as file:
        first_column = []
        second_column = {}
        while True:
            line = file.readline()
            if not line:
                break
            x, y = map(int, line.split())
            first_column.append(x)
            if (not y in second_column):
                second_column[y] = 0
            second_column[y] += 1
        
        Sum = 0
        for x in first_column:
            if x in second_column:
                Sum += x * second_column[x]

        print(Sum)

except:
    print("Error")